Add File watcher support

panaaj opened the pull request:
This PR adds a file watcher to charts folders to detect new files placed in the folder. This makes the new charts available to subsequent requests without having to restart the plugin (#28).

panaaj commented:
@tkurki Can you please review and if all is OK it's good to publish.

This code seemed to solve Freeboard-SK issue https://github.com/SignalK/freeboard-sk/issues/211

tkurki commented on README.md:
````diff
@@ -61,9 +73,46 @@ Publicly available MBTiles charts can be found from:
+For chart files placed in the path(s) defined in the plugin configuration, the url will be:
+
+```bash
+/chart-tiles/${identifier}/${z}/${x}/${y}
````
One of the ideas for using `/signalk` prefix is that one could use a reverse proxy that proxies not only signalk but also other services. This does not hold 100% for the admin ui and other stuff, but I would rather keep this under `/signalk` than at the root.

panaaj replied:
The intent was to not have a chart plugin serve chart tiles under the `/signalk/v2/api/resources/charts` path.
Have shifted the path to be under `/signalk`.

tkurki commented on README.md:
```diff
@@ -61,9 +73,46 @@ Publicly available MBTiles charts can be found from:
+/chart-styles/${mapboxstyle.json}
+
+# when access token is defined
+/chart-styles/${mapboxstyle.json}?access_token=${token}
```
This is local url, right? So why is there an access token needed?

panaaj replied:
The Mapbox style json file is served locally but may reference resources that are remote which require an access token...
_Example: satellite-v9.jon_
```javascript
{
  "version": 8,
  "name": "Mapbox Satellite",
  "metadata": {
    "mapbox:autocomposite": true,
    "mapbox:type": "default"
  },
  "sources": {
    "mapbox": {
      "type": "raster",
      "url": "mapbox://mapbox.satellite",    // remote resource
      "tileSize": 256
    }
  },
...
```

tkurki commented on package.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@signalk/charts-plugin",
-  "version": "3.0.0",
+  "version": "3.1.0",
```
One the the advantages of "changelog from merged PRs" process is that a PR need not (and should not) change the package version, as there is no guarantee in which version a PRs changes are published.

panaaj replied:
reverted.

tkurki commented on src/index.ts:
```diff
@@ -147,80 +164,106 @@ module.exports = (app: ChartProviderApp): Plugin => {
+      `/${chartTilesPath}/:identifier/:z([0-9]*)/:x([0-9]*)/:y([0-9]*)`,
       async (req: Request, res: Response) => {
         const { identifier, z, x, y } = req.params
+        await refreshProviders()
```
To me this style of sprinkling `refreshProviders` seems a bit messy.

I would have implemented this by removing the direct access to `chartProviders`: renaming it to `_chartProviders` or something and introducing `getChartProviders`, where we could have one place to apply the logic of making sure we are serving up to date data.

panaaj replied:
Done.

tkurki commented on src/index.ts:
```diff
@@ -147,80 +164,106 @@ module.exports = (app: ChartProviderApp): Plugin => {
+  const refreshProviders = async () => {
+    const td = Date.now() - (lastWatchEvent as number)
+    app.debug(`last watch event time elapsed = ${td}`)
+    if (lastWatchEvent && td > 5000) {
```
Did I understand the logic correctly: if we notice a change in some chart data we reload the chart data the first time the data is needed after 5 seconds of the change?

Could we not just mark the data as stale and when it is next needed reload it? Or always reload when a change is noticed? Why the 5 second delay?

panaaj replied:
This is due to the way `FSWatcher` produces events and the ability to distinguish when a:
1. New file is created
2. An exisiting file is replaced
3. An existing file is renamed
4. Multiple files are being added / deleted

There are multiple events produced during a simple file rename with no discernable metadata to indicate the actual operation taking place.... so the delay is to ensure the stream of events has ceased to indicate the process has completed. 
Without it we would trigger multiple reloads while a rename, etc. is in process and throw exceptions.

tkurki commented on README.md:
```diff
@@ -2,7 +2,14 @@
 
-Supports both v1 and v2 Signal K resources api paths.
+Chart metadata is derived from the following supported chart files:
+- Mapbox Tiles _(.mbtiles)_
```
Afaik MBTiles is  better known as just [MBTiles](https://github.com/mapbox/mbtiles-spec) and touted as _open format_. The spec README never mentions Mapbox, while it is under their Github org.

We could link to the canonical resources?

panaaj replied:
Updated to MBTiles.
Which canonical resources are you refferring to?

tkurki replied:
The repo I linked.

tkurki commented on src/charts.ts:
```diff
@@ -101,6 +104,36 @@ function openMbtilesFile(file: string, filename: string) {
+  return filename.replace('.json', '').replaceAll(' ', '-').toLocaleLowerCase()
+}
+
+async function openMbstylesFile(file: string, filename: string) {
```
I'd like to rename this to `openMapboxStylesFile`- while the `mb` in `mbtiles` is originally also `MapBox` that format has evolved to be a common name, there is no equivalent `mbstyles`.

panaaj replied:
Renamed to `openMapboxStylesFile`

tkurki commented:
Unrelated to this PR: README is not very consistent
- default location repeated
- the steps-oriented configuration instructions are not very universal: some people will use online charts, others mbtiles etc

panaaj commented on CHANELOG.md:
```diff
@@ -0,0 +1,20 @@
+## Change Log
```
will remove

msallin commented:
I arrived here because I would like to have two things: 1. the file watcher, and 2. support for recursive folders.
I've a share with a folder called "charts," which contains subfolders with regions. Those subfolders contain multiple different charts for each region.

For a new region, I have to create the folder on the share, add it to the plugin, and restart for every chart I add.

This PR implements the watcher. WDYT about letting the user specify recursive paths?

msallin committed:
feat: watch chart folders and reload on file changes

Adds an fs.watch on each configured chart path so new, renamed or deleted
chart files are picked up without a plugin restart. Watcher bursts during
rename or atomic-save sequences are collapsed into a single reload via a
setTimeout debounce (5s default, overridable via SK_CHARTS_RELOAD_DEBOUNCE_MS
for tests), which also prevents parallel loadCharts() runs from stepping
on each other.

Chart discovery is recursive: directories without tilemapresource.xml or
metadata.json are descended into, so layouts like charts/<region>/<chart>
work without listing every subdir in the config. Depth is bounded and
symlinks are skipped so a misplaced chart path can't walk into node_modules
or loop. fs.watch is opened with recursive:true (Node 22+ required anyway)
and falls back to top-level watching if the platform rejects it.

On reload, MBTiles SQLite connections are reconciled: handles for files
still present are reused instead of being replaced, and handles for files
that have gone away are closed. Without this, each watcher-triggered
reload leaked one connection per .mbtiles file. stop() also closes the
connections so the user can move or delete files while the plugin is
stopped (Windows blocks deletion on open handles).

A catastrophic reload failure now keeps the last-good chartProviders
rather than blanking the service out. Plugin status reflects the current
chart count after every successful load.

Test teardown calls plugin.stop() so watchers and handles don't keep the
event loop alive after mocha finishes.

Closes #28
